File: totp_implementation/totp.py
import hashlib
import time

def totp(secret: bytes, hfunc = hashlib.sha1, X: int = 30, T0: int = 0, step: int = None) -> int:
    if step is None:
        cut = int(time.time())
        step = (cut - T0) // X
    return hotp(secret, step, hfunc=hfunc)

def hotp(secret: bytes, counter: bytes | int, digit: int = 6, hfunc = hashlib.sha1) -> int:
    if isinstance(counter, int):
        counter = counter.to_bytes(8, byteorder='big')
    elif len(counter) != 8:
        raise TypeError('Counter must be 8 bytes.')
    
    hs = hmac(secret, counter, hfunc)
    offset = hs[-1] & 0xf
    p = int.from_bytes(hs[offset : offset + 4], 'big')
    return (p & 0x7fffffff) % (10 ** digit)

def hmac(key: bytes, msg: bytes, hfunc):
    def xor(a, b):
        return bytes([aa ^ bb for aa, bb in zip(a, b)])
    
    B = hfunc().block_size
    ipad = b'\x36' * B
    opad = b'\x5c' * B
    
    if len(key) > B:
        key = hfunc(key).digest()
    kpad = key + b'\x00' * (B - len(key))

    k_opad = xor(kpad, opad)
    k_ipad = xor(kpad, ipad)
    return hfunc(k_opad + hfunc(k_ipad + msg).digest()).digest()

File: totp_implementation/totp.c
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <openssl/sha.h>

#define B 64
#define B_8 (B + 8)
#define B_20 (B + 20)
#define SECRET_LENGTH 20
#define SHA_DIGEST_LENGTH_m1 19

#define X 30
#define T0 0

void xor(unsigned char *a, unsigned char *res, int length) {
    for (int i = 0; i < length; i++) {
        res[i] ^= a[i];
    }
}

void hmac(unsigned char *key, unsigned char *msg, unsigned char *mac) {
    int i;
    unsigned char ipad[B];
    unsigned char opad[B];
    unsigned char kipad[B] = {0};
    unsigned char kopad[B] = {0};
    unsigned char k_ipad_msg[B_8];
    unsigned char k_opad_digest[B_20];
    unsigned char digest[SHA_DIGEST_LENGTH];
    SHA_CTX ctx;
    
    memset(ipad, '\x36', B);
    memset(opad, '\x5c', B);

    // kipad = K ^ ipad
    memcpy(kipad, key, SECRET_LENGTH);
    xor(ipad, kipad, B);

    // kopad = K ^ opad
    memcpy(kopad, key, SECRET_LENGTH);
    xor(opad, kopad, B);

    // k_ipad_msg = (K ^ ipad) || msg
    memcpy(k_ipad_msg, kipad, B);
    memcpy(&k_ipad_msg[B], msg, 8);

    // digest = SHA1((K ^ ipad) || msg)
    SHA1_Init(&ctx);
    SHA1_Update(&ctx, k_ipad_msg, B_8);
    SHA1_Final(digest, &ctx);

    // k_opad_digest = (K ^ opad) || SHA1((K ^ ipad) || msg)
    memcpy(k_opad_digest, kopad, B);
    memcpy(&k_opad_digest[B], digest, SHA_DIGEST_LENGTH);

    // mac = SHA1((K ^ opad) || SHA1((K ^ ipad) || msg))
    SHA1_Init(&ctx);
    SHA1_Update(&ctx, k_opad_digest, B_20);
    SHA1_Final(mac, &ctx);
}

int hotp(unsigned char *secret, unsigned char *step) {
    unsigned char hs[SHA_DIGEST_LENGTH];
    int offset;
    unsigned int p;
    int i;

    hmac(secret, step, hs);
    offset = hs[SHA_DIGEST_LENGTH_m1] & 0xf;

    p = 0;
    for (i = 0; i < 4; i++) p += (unsigned int)hs[offset + i] << (24 - i * 8);

    return (p & 0x7fffffff) % 1000000;
}

int totp(unsigned char *secret) {
    time_t cut;
    unsigned long step_long;
    unsigned char step[8];
    int i;

    cut = time(NULL);
    step_long = (cut - T0) / X;
    for (i = 0; i < 8; i++) step[i] = (step_long >> (56 - i * 8)) & 0xff;

    return hotp(secret, step);
}